File: src/main/java/com/soze/r2d/UiState.java
package com.soze.r2d;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class UiState {

  private final Map<String, Object> state = new HashMap<>();

  public UiState() {

  }

  public UiState(Map<String, Object> state) {
    this.state.clear();
    this.state.putAll(state);
  }

  public Object get(String key) {
    return state.get(key);
  }

  public <T> T get(String key, T defaultValue) {
    return (T) state.getOrDefault(key, defaultValue);
  }

  public void set(String key, Object value) {
    state.put(key, value);
  }

  public void merge(UiState uiState) {
    for (Map.Entry<String, Object> entry : uiState.entries()) {
      this.state.put(entry.getKey(), entry.getValue());
    }
  }

  private Set<Map.Entry<String, Object>> entries() {
    return state.entrySet();
  }

  public void clear() {
    this.state.clear();
  }

}

File: src/main/kotlin/com/soze/r2d/UiState.kt
package com.soze.lifegame.uibuilder

data class UiState(
	val state: MutableMap<String, Any?> = HashMap()
) {

	constructor(vararg nextState: Pair<String, Any?>) : this() {
		nextState.forEach {
			state[it.first] = it.second
		}
	}

	operator fun get(key: String): Any? {
		return state[key]
	}

	operator fun set(key: String, value: Any?) {
		state[key] = value
	}

	private fun entries(): Set<Map.Entry<String, Any?>> {
		return state.entries
	}

	fun merge(uiState: UiState) {
		uiState.entries().forEach {
			set(it.key, it.value)
		}
	}

	fun merge(vararg nextState: Pair<String, Any?>) {
		nextState.forEach {
			state[it.first] = it.second
		}
	}

	fun clear() {
		state.clear()
	}

}